File: 1 SEMESTRE/EXERCICIOS DE LISTA, PILHA E FILA/Lista de Pilhas/Exercicios em Java/Exercicio3.java
/*Considere uma lista de pilhas, e crie o método int somaMaiorPilha(), que retorne o nó da lista, com a maior soma dos elementos da pilha 
JAVA*/

class Lista {
    CelulaLista inicio; 
    CelulaLista fim; 
}

class CelulaLista {
    CelulaPilha topo; 
    CelulaLista prox; 
}

class CelulaPilha {
    int elemento; 
    CelulaPilha prox; 
}

//Função para retornar um inteiro, que é a soma da maior Pilha. 
int somaMaiorPilha() {
    //Variável para guardar a pilha com maior soma de seus elementos. 
    int maiorSoma = 0; 

    //Percorrer a lista por completo. 
    for (CelulaLista i = inicio; i != null; i = i.prox) {
        //Variável para somar os elementos da pilha. 
        int somaAtual = 0; 

        //Percorrer a pilha por completo e somar os elementos dela. 
        for (CelulaPilha j = i.topo; j != null; j = j.prox) {
            somaAtual += j.elemento; 
        }

        //Verificar se a soma atual dos elementos é maior que a soma total. 
        if (somaAtual > maiorSoma) {
            maiorSoma = somaAtual; 
        }
    }

    //Retornando inteiro e não célula. 
    return maiorSoma; 
}

File: 1 SEMESTRE/EXERCICIOS DE LISTA, PILHA E FILA/Lista de Pilhas/Exercicios em C/Exercicio3.c
/*Mesmo exercício de lista de pilhas, referente ao exercício 3, porém agora na linguagem C. */

#include <stdio.h> 
#include <stdlib.h> 

typedef struct Lista {
    CelulaLista *inicio; 
    CelulaLista *fim; 
} Lista; 

typedef struct CelulaLista {
    CelulaPilha *topo; 
    CelulaLista *prox; 
} CelulaLista; 

typedef struct CelulaPilha {
    int elemento; 
    CelulaLista *prox; 
} CelulaPilha; 

//Função que retorna um inteiro
int somaMaiorPilha (Lista *lista) {
    //Criar variável para guardar a soma. 
    int maiorSoma = 0; 

    //Percorrer a lista 
    for (CelulaLista *i = lista->inicio; i != NULL; i = i->prox) {
        //Variável para a soma dos elementos da pilha. 
        int somaAtual = 0; 

        //Percorrer a pilha e ir somando os elementos. 
        for (CelulaPilha *j = i->topo; j != NULL; j = j->prox) {
            somaAtual += j->elemento; 
        }

        //Verificar a soma
        if (somaAtual > maiorSoma) {
            maiorSoma = somaAtual; 
        }
    }

    //Retornar a maior soma
    return maiorSoma; 
}